### PR TITLE
parameter secret_token

### PR DESCRIPTION
**Pull request by sheldygg**
# Description

Added X-Telegram-Bot-Api-Secret-Token header check

- [x] New feature (non-breaking change which adds functionality)

**Test Configuration**:
* Operating System: Windows 10
* Python version: 3.11

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: aiogram/webhook/aiohttp_server.py [hunk 1]
```diff
@@ -184,6 +186,10 @@ async def _handle_request(self, bot: Bot, request: web.Request) -> web.Response:
         return web.Response(body=self._build_response_writer(bot=bot, result=result))
 
     async def handle(self, request: web.Request) -> web.Response:
+        if self.secret_token:
+            t_token = request.headers.get("X-Telegram-Bot-Api-Secret-Token")
```
Comment (JrooTJunior) [inline aiogram/webhook/aiohttp_server.py:190]: What does mean prefix `t_`?

I any short prefixes is used in project it should be conventioned in the contribution guide, but as you see we doesn't have any prefixes convention in the project.

So, please set correct and understandable name for the variable.

**Review by JrooTJunior — CHANGES_REQUESTED** [review]
1. Changes is not covered by tests
2. Is not documented
3. Can be applied only for simple static secret token validation (`SimpleRequestHandler`-like handlers wtih static bot instance)
4. Is totally not compatible with multibot-systems

So, this feature should be discussed before implementation

**Comment by JrooTJunior** [comment]
Any news?

**Comment by sheldygg** [comment]
> 

unban in tg chat. I dont quite understand what I need to do

**Comment by JrooTJunior** [comment]
It is unacceptable to justify Russian terrorism and violation of the territorial integrity of Ukraine, I don't want this to happen in my chats, that's why you can't be unbanned.

**Comment by sheldygg** [comment]
I did not justify, I saw a person who wrote something in the chat, and it was written in his channel that the LPR is Russia. I wanted him banned and threw it into the chat. Your wife did it, but then you came and banned me. I don't support Russians, I'm from Ukraine, Chernihiv region

**Comment by sheldygg** [comment]
t.me/kugona
profile of the person who wrote it

**Comment by sheldygg** [comment]
я не виноват =(

**Comment by JrooTJunior** [comment]
But you send the same message

**Comment by sheldygg** [comment]
> But you send the same message

I wanted everyone to see what he had in his channel.  those who communicated at that time in the chat understood what i mean

**Comment by JrooTJunior** [comment]
Glory to the heroes!

Ok, you are unlocked, but you can also visit the Ukrainian chat.

File: aiogram/webhook/aiohttp_server.py [hunk 2]
```diff
@@ -131,6 +133,10 @@ async def resolve_bot(self, request: web.Request) -> Bot:
         """
         pass
 
+    @abstractmethod
+    def secret_verify(self, secret_token: str) -> bool:
```
Comment (JrooTJunior) [inline aiogram/webhook/aiohttp_server.py:137]: please change these words (a verb followed by a noun - means do "something" on "thing")

```suggestion
    def veryfy_secret(self, secret_token: str) -> bool:
```

File: aiogram/webhook/aiohttp_server.py [hunk 3]
```diff
@@ -184,6 +190,8 @@ async def _handle_request(self, bot: Bot, request: web.Request) -> web.Response:
         return web.Response(body=self._build_response_writer(bot=bot, result=result))
 
     async def handle(self, request: web.Request) -> web.Response:
+        if not self.secret_verify(request.headers.get("X-Telegram-Bot-Api-Secret-Token")):
```
Comment (JrooTJunior) [inline aiogram/webhook/aiohttp_server.py:193]: I believe this check should be performed after the bot instance is already known

This is required to check the secret for a specific bot instance

File: aiogram/webhook/aiohttp_server.py [hunk 4]
```diff
@@ -90,6 +90,7 @@ def __init__(
         self,
         dispatcher: Dispatcher,
         handle_in_background: bool = False,
+        secret_token: Optional[str] = None,
```
Comment (JrooTJunior) [inline aiogram/webhook/aiohttp_server.py:93]: This argument is really needed in base request handler?

Is unsuded in base class but only needed in simple reuqest handler.

**Review by JrooTJunior — CHANGES_REQUESTED** [review]

File: aiogram/webhook/aiohttp_server.py [hunk 5]
```diff
@@ -198,16 +204,27 @@ class SimpleRequestHandler(BaseRequestHandler):
...
-        super().__init__(dispatcher=dispatcher, handle_in_background=handle_in_background, **data)
+        super().__init__(
+            dispatcher=dispatcher,
+            handle_in_background=handle_in_background,
+            **data)
         self.bot = bot
+        self.secret_token = secret_token
+
+    def verify_secret(self, telegram_secret_token: str) -> bool:
+        return telegram_secret_token == self.secret_token
```
Comment (Olegt0rr) [inline aiogram/webhook/aiohttp_server.py:227]: It's a bad practice to compare secrets via `==`. Use [compare digest](https://docs.python.org/3/library/secrets.html#secrets.compare_digest)

**Review by Olegt0rr — CHANGES_REQUESTED** [review]

File: aiogram/webhook/aiohttp_server.py [hunk 6]
```diff
@@ -198,16 +204,27 @@ class SimpleRequestHandler(BaseRequestHandler):
     """
 
     def __init__(
-        self, dispatcher: Dispatcher, bot: Bot, handle_in_background: bool = True, **data: Any
+        self, dispatcher: Dispatcher,
```
Comment (Olegt0rr) [inline aiogram/webhook/aiohttp_server.py:207]:
```suggestion
        self, 
        dispatcher: Dispatcher,
```

File: aiogram/webhook/aiohttp_server.py [hunk 7]
```diff
@@ -185,6 +189,8 @@ async def _handle_request(self, bot: Bot, request: web.Request) -> web.Response:
 
     async def handle(self, request: web.Request) -> web.Response:
         bot = await self.resolve_bot(request)
+        if not self.verify_secret(request.headers.get("X-Telegram-Bot-Api-Secret-Token")):
+            return web.Response(body="Unauthorized", status=401)
```
Comment (Olegt0rr) [inline aiogram/webhook/aiohttp_server.py:193]: So what if I don't wanna verify secrets?

Comment (sheldygg) [reply aiogram/webhook/aiohttp_server.py:193]: i dont have ideas :(

Comment (JrooTJunior) [reply aiogram/webhook/aiohttp_server.py:193]: You should return `True` by default from `verify_secret` method

Comment (Olegt0rr) [reply aiogram/webhook/aiohttp_server.py:193]: I think we should make a single choice of implementation level.
- `self.verify_secret` implemented on base level
- `self.secret_token` implemented on upper level

So, we can push down secret token and check it's not None before trying to get headers and verify something.

**Comment by sheldygg** [comment]
hm...

**Review by JrooTJunior — APPROVED** [review]

**Pull request merged**